express_server.js

```diff
@@ -6,135 +6,175 @@ const cookieSession = require("cookie-session");
 const bcrypt = require("bcrypt");
 const methodOverride = require("method-override");
 // helper functions:
-const { generateID, getUserByEmail, userURLs, copy } = require("./helpers.js");
+const {
+  areValidEmailPassword,
+  generateID,
+  getUserByEmail,
+  hashPassword,
+  userURLs,
+  emailExists,
+} = require("./helpers.js");
 
 app.set("view engine", "ejs");
-app.use(bodyParser.urlencoded({extended: true}));
-app.use(cookieSession({
-  name: "session",
-  keys: ["user_id"]
-}));
-app.use(methodOverride('_method'));
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(
+  cookieSession({
+    name: "session",
+    keys: ["user_id"],
+  }),
+);
+app.use(methodOverride("_method"));
 
 // used to keep track of all the URLs and their shortened forms
 const urlDatabase = {
-  "b2xVn2": { longURL: "http://www.lighthouselabs.ca", userID: "dummy" },
-  "icon": { longURL: "https://i.pinimg.com/originals/b9/3f/80/b93f8070c4178d1906198e264cb4c98f.png", userID: "dummy" }
+  b2xVn2: { longURL: "http://www.lighthouselabs.ca", userID: "dummy" },
+  icon: {
+    longURL:
+      "https://i.pinimg.com/originals/b9/3f/80/b93f8070c4178d1906198e264cb4c98f.png",
+    userID: "dummy",
+  },
 };
 
 //global variable, used to store and access users in the app
-let users = {  };
+const users = {};
+
+// Create an endpoint that renders an error info page.
+app.get("/errors/:statusCode", (req, res) => {
+  templateVars = {
+    statusCode: req.params.statusCode,
+    user: req.session.id,
+  };
+  res.render("url_errors", templateVars);
+});
 
 //registers the user (handles registration form data)
-app.post("/register", function(req, res) {
-  const newUserID = generateID();
+app.post("/register", (req, res) => {
   if (!req.body.email || !req.body.password) {
     // errors if fields are empty
-    res.status(400).send("Oops! Please enter an email and password :) <a href='/urls/'>Register</a>");
+    res.status(406).redirect("/errors/406");
   } else if (getUserByEmail(req.body.email, users)) {
-    res.send("Hey! We're already friends, just <a href='/login'>log in!</a> :)");
+    res.status(409).redirect("/errors/409");
   } else {
-    users[newUserID] = {id: newUserID, email: req.body.email, password: bcrypt.hashSync(req.body.password, 10)};
+    const newUserID = generateID();
+    users[newUserID] = {
+      id: newUserID,
+      email: req.body.email,
+      password: hashPassword(req.body.password),
+    };
     req.session.user_id = newUserID;
     res.redirect("/urls");
   }
 });
 
 //renders register template
-app.get('/register', function(req, res) {
-  let templateVars = { urls: urlDatabase, user:users[req.session.user_id] };
+app.get("/register", (req, res) => {
+  const templateVars = { user: req.session.user_id };
```

Comment:
No need to pass the urls here nor the specific user object. Your view only needs to know if there's a valid user session so it can decide to render this or that. Always check your code for things not needed like these.